### PR TITLE
feat: implement datafusion tableprovider trait for `Dataset`

### PR DESCRIPTION
**Pull request by universalmind303**
This allows better native integrations with datafusion.
closes https://github.com/lancedb/lance/issues/1520

### Example

```rust
use std::sync::Arc;

use datafusion::prelude::*;
use lance::Dataset;

#[tokio::main]
async fn main() -> datafusion::error::Result<()> {
    let path = "./.lancedb/my_table.lance";

    let dataset = Dataset::open(path).await.unwrap();
    let ctx = SessionContext::new();
    ctx.register_table("my_table", Arc::new(dataset)).unwrap();
    let df = ctx
        .sql("SELECT * FROM my_table where lat = 45.5")
        .await
        .unwrap();

    df.show().await?;
    Ok(())
}
```

### CODE REVIEW

**Comment by universalmind303** [comment]
@wjones127 I didn't realize that there was already an issue open for this until just now. I think there's a bit more work that could be done with pushdowns (such as partitioning). But this PR serves as a pretty good starting point that supports  projection, predicate & slice pushdowns.

File: rust/lance/src/datafusion/logical_plan.rs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+    fn get_table_definition(&self) -> Option<&str> {
+        None
+    }
+
+    fn get_logical_plan(&self) -> Option<&LogicalPlan> {
+        None
+    }
+
+    async fn scan(
```
Comment (universalmind303) [inline rust/lance/src/datafusion/logical_plan.rs:43]: Originally I did the same as was done here https://github.com/lancedb/lance/pull/1521/files#diff-aa127fed9fd03bbd48ba671543f8fd2991985b373cca51bf620ae9ca264810caR47-R81, but it required a bit of extra processing to go through the scanner. So i refactored it to use the same logic as the scanner. This way avoids unnecessary allocations & compute.

File: rust/lance/src/datafusion/logical_plan.rs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            let filter_schema = Arc::new(scanner.dataset.schema().project(&columns_in_filter)?);
+            scanner.scan(true, false, filter_schema)
+        } else {
+            // Scan without filter or limits
+            scanner.scan(
+                scanner.with_row_id,
+                false,
+                scanner.projections.clone().into(),
+            )
+        };
```
Comment (wjones127) [inline rust/lance/src/datafusion/logical_plan.rs:81]: I think we can ignore a bunch of options that aren't used:

```suggestion
        if let Some(expr) = filter_expr.as_ref() {
            let columns_in_filter = column_names_in_expr(expr.as_ref());
            let filter_schema = Arc::new(scanner.dataset.schema().project(&columns_in_filter)?);
            scanner.scan(true, false, filter_schema)
        } else {
            // Scan without filter or limits
            scanner.scan(
                false,
                false,
                scanner.projections.clone().into(),
            )
        };
```

File: rust/lance/src/datafusion/logical_plan.rs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                        columns_in_filter.push(column.column_name.clone())
+                    }
+                }
+            }
+            let filter_schema = Arc::new(scanner.dataset.schema().project(&columns_in_filter)?);
+            let remaining_schema = filter_schema.exclude(plan.schema().as_ref())?;
+            if !remaining_schema.fields.is_empty() {
+                // Not all columns for filter are ready, so we need to take them first
+                plan = scanner.take(plan, &remaining_schema, scanner.batch_readahead)?;
+            }
```
Comment (wjones127) [inline rust/lance/src/datafusion/logical_plan.rs:100]: None of this is relevant. If we have a filter, then we already pulled all the columns in the original scan. And we aren't doing any sorting in this node, so we can ignore that part too.
```suggestion
```

File: rust/lance/src/datafusion/logical_plan.rs
Comment (wjones127) [inline rust/lance/src/datafusion/logical_plan.rs:1]: Could you add a test like the example you have in the PR description?

File: rust/lance/src/dataset/scanner.rs [hunk 4]
```diff
@@ -124,27 +124,27 @@ impl ColumnOrdering {
 ///   .sum()
 /// ```
 pub struct Scanner {
-    dataset: Arc<Dataset>,
+    pub(crate) dataset: Arc<Dataset>,
```
Comment (wjones127) [inline rust/lance/src/dataset/scanner.rs:127]: I'm not sure we want to make these `pub(crate)`. Perhaps we should move the `TableProvider` implementation into this file, or a submodule of it?

**Review by wjones127 — CHANGES_REQUESTED** [review]
Thanks for making this PR. As you noticed, I was just starting to pursue the same thing. I made some suggestions to make the changes a cleaner.

**Comment by universalmind303** [comment]
Thanks for the feedback @wjones127. I realized i was greatly over complicating the `scan` function. This is a much thinner implementation now. Without a custom physical optimizer rule, datafusion wont fully pushdown the limits & predicates, so I left the handling of those to datafusion.

The limit still gets pushed down to the `Scanner`, but without custom rules, datafusion will still wrap it in a `GlobalLimitExec`. 

We could probably expose a Lance specific optimizer rule that can be plugged in to datafusion to handle more pushdowns, but that is beyond the scope of this PR.

**Comment by wjones127** [comment]
> Without a custom physical optimizer rule, datafusion wont fully pushdown the limits & predicates, so I left the handling of the predicates to datafusion. The limit still gets pushed down to the Scanner, but without custom rules, datafusion will still wrap it in a GlobalLimitExec. Not a big deal, but it would be preferred to drop that extra node as it's already pushed down to the Scanner.

> We could probably expose a Lance specific optimizer rule that can be plugged in to datafusion to handle more pushdowns, but that is beyond the scope of this PR.

Thanks, that's helpful feedback. We haven't looked too far into integrating with the DataFusion optimizer, but it's something we'd eventually like to do. I think we first have some work to make our SQL parser work, since we customize that.

File: rust/lance-core/src/io/reader.rs
Comment (wjones127) [inline rust/lance-core/src/io/reader.rs:1]: What are these changes for? They seem unrelated.

File: rust/lance-core/src/io/reader.rs [hunk 5]
```diff
@@ -898,13 +898,19 @@ async fn read_struct_array(
...
-        sub_arrays.push((Arc::new(child.into()), arr));
+        fields.push(child.into());
+
+        arrays.push(arr);
     }
 
-    Ok(Arc::new(StructArray::from(sub_arrays)))
+    let sa = StructArray::try_new(fields.into(), arrays, None)?;
+
+    Ok(Arc::new(sa))
```
Comment (universalmind303) [inline rust/lance-core/src/io/reader.rs:913]: I was getting some panics while debugging that were hard to track down. Since `from` uses `try_new` under the hood, it makes more sense to bubble up the error instead of potentially panicking.

File: rust/lance/src/datafusion/logical_plan.rs [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+        Dataset::write(batch_reader, test_uri, Some(WriteParams::default()))
+            .await
+            .unwrap();
+
+        let dataset = Dataset::open(test_uri).await.unwrap();
+        let ctx = SessionContext::new();
+        ctx.register_table("my_table", Arc::new(dataset)).unwrap();
+        let df = ctx
+            .sql("SELECT point.lat as lat, point.long as long FROM my_table")
```
Comment (wjones127) [inline rust/lance/src/datafusion/logical_plan.rs:192]: Do you know what expression type a struct field reference becomes? Been trying to figure that out.

File: rust/lance-core/src/io/reader.rs
Comment (universalmind303) [reply rust/lance-core/src/io/reader.rs:1]: They were for initially debugging a panic with a schema mismatch. See other comment. I can revert these if you prefer.

File: rust/lance/src/datafusion/logical_plan.rs (hunk 6 above)
Comment (wjones127) [reply rust/lance/src/datafusion/logical_plan.rs:192]: I guess GetIndexedField https://docs.rs/datafusion/latest/datafusion/logical_expr/struct.GetIndexedField.html

Comment (universalmind303) [reply rust/lance/src/datafusion/logical_plan.rs:192]: Yeah it'll be `GetIndexedField`

```sql
select point.long from my_table
```
will get evaluated to this expr

```
Expr::GetIndexedField(
    GetIndexedField {
        expr: Column(
            Column {
                relation: Some(
                    Bare {
                        table: "my_table",
                    },
                ),
                name: "point",
            },
        ),
        field: NamedStructField {
            name: Utf8("long"),
        },
    },
)
```

**Review by wjones127 — APPROVED** [review]
